tokio-sync/src/task/atomic_task.rs

```diff
@@ -0,0 +1,292 @@
+use ::loom::{
+    futures::task::{self, Task},
+    sync::CausalCell,
+    sync::atomic::AtomicUsize,
+};
+
+use std::fmt;
+use std::sync::atomic::Ordering::{Acquire, Release, AcqRel};
+
+/// A synchronization primitive for task notification.
+///
+/// `AtomicTask` will coordinate concurrent notifications with the consumer
+/// potentially "updating" the underlying task to notify. This is useful in
+/// scenarios where a computation completes in another thread and wants to
+/// notify the consumer, but the consumer is in the process of being migrated to
+/// a new logical task.
+///
+/// Consumers should call `register` before checking the result of a computation
+/// and producers should call `notify` after producing the computation (this
+/// differs from the usual `thread::park` pattern). It is also permitted for
+/// `notify` to be called **before** `register`. This results in a no-op.
+///
+/// A single `AtomicTask` may be reused for any number of calls to `register` or
+/// `notify`.
+///
+/// `AtomicTask` does not provide any memory ordering guarantees, as such the
+/// user should use caution and use other synchronization primitives to guard
+/// the result of the underlying computation.
+pub struct AtomicTask {
```

Comment:
How does this type differ from `loom::futures::AtomicTask`?

Reply:
Oh, I don't really know what I'm talking about. I was just looking at `Semaphore` and saw the import `use loom::futures::AtomicTask`, and wondered what the difference was between that one and this one.